ciphersuite/ed25519.go

```diff
@@ -0,0 +1,221 @@
+func NewEd25519CipherSuite() *Ed25519CipherSuite {
+	return &Ed25519CipherSuite{}
+}
+
+// Name returns the name of the suite.
+func (s *Ed25519CipherSuite) Name() Name {
+	return Ed25519CipherSuiteName
+}
+
+// PublicKey takes a raw public key and converts it to a public key.
+func (s *Ed25519CipherSuite) PublicKey(raw *RawPublicKey) (PublicKey, error) {
+	if raw.Name() != s.Name() {
+		return nil, xerrors.New(errNotEd25519CipherSuite)
+	}
+
+	if len(raw.Data) != ed25519.PublicKeySize {
+		return nil, xerrors.New(errInvalidBufferSize)
+	}
+
+	return &Ed25519PublicKey{data: raw.Data}, nil
+}
+
+// SecretKey takes a raw secret key and converts it to a secret key.
+func (s *Ed25519CipherSuite) SecretKey(raw *RawSecretKey) (SecretKey, error) {
+	if raw.Name() != s.Name() {
+		return nil, xerrors.New(errNotEd25519CipherSuite)
+	}
+
+	if len(raw.Data) != ed25519.PrivateKeySize {
+		return nil, xerrors.New(errInvalidBufferSize)
+	}
+
+	return &Ed25519SecretKey{data: raw.Data}, nil
+}
+
+// Signature takes a raw signature and converts it to a signature.
+func (s *Ed25519CipherSuite) Signature(raw *RawSignature) (Signature, error) {
+	if raw.Name() != s.Name() {
+		return nil, xerrors.New(errNotEd25519CipherSuite)
+	}
+
+	if len(raw.Data) != ed25519.SignatureSize {
+		return nil, xerrors.New(errInvalidBufferSize)
+	}
+
+	return &Ed25519Signature{data: raw.Data}, nil
+}
+
+// GenerateKeyPair generates a secret key and its associated public key. When
+// reader is nil, it will use the default randomness source.
+func (s *Ed25519CipherSuite) GenerateKeyPair(reader io.Reader) (PublicKey, SecretKey, error) {
+	pk, sk, err := ed25519.GenerateKey(reader)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("generate ed25519 key: %v", err)
+	}
+
+	return &Ed25519PublicKey{data: pk}, &Ed25519SecretKey{data: sk}, nil
+}
+
+// Sign signs the message with the secret key and returns the signature that
+// can be verified with the public key.
+func (s *Ed25519CipherSuite) Sign(sk SecretKey, msg []byte) (Signature, error) {
+	secretKey, err := s.unpackSecretKey(sk)
+	if err != nil {
+		return nil, xerrors.Errorf("unpacking secret key: %v", err)
+	}
+	sigbuf := ed25519.Sign(secretKey.data, msg)
+
+	return &Ed25519Signature{data: sigbuf}, nil
+}
+
+// Verify returns nil when the signature of the message can be verified by
+// the public key.
+func (s *Ed25519CipherSuite) Verify(pk PublicKey, sig Signature, msg []byte) error {
+	publicKey, err := s.unpackPublicKey(pk)
+	if err != nil {
+		return xerrors.Errorf("unpacking public key: %v", err)
+	}
+
+	signature, err := s.unpackSignature(sig)
+	if err != nil {
+		return xerrors.Errorf("unpacking signature: %v", err)
+	}
+
+	if !ed25519.Verify(publicKey.data, msg, signature.data) {
+		return xerrors.New("signature not verified")
+	}
+
+	return nil
+}
+
+func (s *Ed25519CipherSuite) unpackPublicKey(pk PublicKey) (*Ed25519PublicKey, error) {
+	if data, ok := pk.(*RawPublicKey); ok {
```

Comment:
Why the casting? That looks like try to cope with library misuse; simply fail looks right

Reply:
and related below

Reply:
Because at this moment we don't use network messages and data structures in a separate way so I made the raw elements to be accepted as valid abstraction. The cipher suite must then convert them. It will very likely change when we will switch to the official protobuf.

Reply:
TODO it or log it then?